File: data_transfer_demo/http_request/import_data_to_google_sheet.py
import requests
import time
import hashlib
import hmac
import json
import pandas as pd
import os
import pickle
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request

api_key = 'xxxxxxxxxxxxxxx'
secret_key = 'xxxxxxxxxxxxxxxxxxxxxx'
httpClient = requests.Session()
recv_window = str(5000)
url = "https://api-testnet.bybit.com"  # Testnet endpoint
SCOPES = ['https://www.googleapis.com/auth/spreadsheets']

def get_google_auth_token():
    creds = None

    # The file token.pickle stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first time.
    if os.path.exists('token.pickle'):
        with open('token.pickle', 'rb') as token:
            creds = pickle.load(token)

    # If there are no (valid) credentials available, prompt the user to log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                'credentials.json', SCOPES)
            creds = flow.run_local_server(port=0)
            # Save the credentials for the next run
            with open('token.pickle', 'wb') as token:
                pickle.dump(creds, token)

    return creds.token
def HTTP_Request(endPoint, method, payload, Info):
    global time_stamp
    time_stamp = str(int(time.time() * 10 ** 3))
    signature = genSignature(payload)
    headers = {
        'X-BAPI-API-KEY': api_key,
        'X-BAPI-SIGN': signature,
        'X-BAPI-SIGN-TYPE': '2',
        'X-BAPI-TIMESTAMP': time_stamp,
        'X-BAPI-RECV-WINDOW': recv_window,
        'Content-Type': 'application/json'
    }
    if (method == "POST"):
        response = httpClient.request(method, url + endPoint, headers=headers, data=payload)
    else:
        response = httpClient.request(method, url + endPoint + "?" + payload, headers=headers)
    print(response.text)
    print(Info + " Elapsed Time : " + str(response.elapsed))

    response_json = response.json()
    list_data = response_json['result']['list']
    df = pd.DataFrame(list_data)
    values_to_append = df.values.tolist()
    spreadSheetId = "xxxxxxxxxxxx"  # replace by your sheet id
    append_to_google_sheet(spreadSheetId, "A1", values_to_append)

def genSignature(payload):
    param_str = str(time_stamp) + api_key + recv_window + payload
    hash = hmac.new(bytes(secret_key, "utf-8"), param_str.encode("utf-8"), hashlib.sha256)
    signature = hash.hexdigest()
    return signature


def append_to_google_sheet(spreadsheet_id, range_, values):
    # Define the endpoint URL (replace YOUR_SPREADSHEET_ID and YOUR_RANGE with appropriate values)
    endpoint = f"https://sheets.googleapis.com/v4/spreadsheets/{spreadsheet_id}/values/{range_}:append"

    # The access token you got after OAuth2.0 authorization
    access_token = get_google_auth_token()
    headers = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json"
    }
    # Prepare the request data
    body = {
        "values": values  # This should be a list of lists representing rows of your data
    }
    params = {
        "valueInputOption": "RAW",
        "insertDataOption": "INSERT_ROWS"
    }
    response = requests.post(endpoint, headers=headers, params=params, data=json.dumps(body))
    if response.status_code == 200:
        print("Data appended successfully")
    else:
        print("Failed to append data:", response.text)


# Get unfilled wallet balance
endpoint = "/v5/market/tickers"
method = "GET"
params = 'category=linear'
HTTP_Request(endpoint, method, params, "Market Ticker Data")

File: data_transfer_demo/http_request/import_data_to_google_sheet.java
package com.bybit.api.utils;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.extensions.java6.auth.oauth2.AuthorizationCodeInstalledApp;
import com.google.api.client.extensions.jetty.auth.oauth2.LocalServerReceiver;
import com.google.api.client.googleapis.auth.oauth2.GoogleAuthorizationCodeFlow;
import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.javanet.NetHttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.gson.GsonFactory;
import com.google.api.client.util.store.FileDataStoreFactory;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import com.google.api.services.sheets.v4.Sheets;
import com.google.api.services.sheets.v4.SheetsScopes;
import com.google.api.services.sheets.v4.model.*;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

public class import_data_to_google_sheet {
    private static final String API_KEY = "xxxxxxxxx";
    private static final String SECRET_KEY = "xxxxxxxxxxxxxxxxxxxxxx";
    private static final String URL = "https://api-testnet.bybit.com";
    private static final OkHttpClient httpClient = new OkHttpClient();
    private static final ObjectMapper objectMapper = new ObjectMapper();
    private static final String spreadSheetId = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";
    public static void main(String[] args) throws GeneralSecurityException, IOException {
        getSheetsService();
        String endpoint = "/v5/market/tickers";
        String params = "category=linear";
        var data = HttpRequestData(endpoint, "GET", params, "Market Ticker Data");
        String range = "A1"; // Or any appropriate range
        exportToSheet(Objects.requireNonNull(data), spreadSheetId, range);
    }

    private static Sheets getSheetsService() throws IOException, GeneralSecurityException {
        NetHttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
        JsonFactory jsonFactory = GsonFactory.getDefaultInstance();
        Credential credential = getCredentials(httpTransport);

        return new Sheets.Builder(httpTransport, jsonFactory, credential)
                .setApplicationName("TestGoogleSheetApi")
                .build();
    }

    public static void exportToSheet(JsonNode listData, String spreadsheetId, String range) throws IOException, GeneralSecurityException {
        List<List<Object>> values = new ArrayList<>();

        JsonNode firstItem = listData.get(0);
        List<Object> headers = new ArrayList<>();
        firstItem.fieldNames().forEachRemaining(headers::add);

        values.add(headers); // Adding header to values

        for (int i = 0; i < listData.size(); i++) {
            List<Object> row = new ArrayList<>();
            JsonNode item = listData.get(i);
            for (Object header : headers) {
                row.add(item.get(header.toString()).asText());
            }
            values.add(row);  // Adding each row of data to values
        }

        ValueRange body = new ValueRange().setValues(values);
        AppendValuesResponse result = getSheetsService().spreadsheets().values()
                .append(spreadsheetId, range, body)
                .setValueInputOption("RAW")
                .execute();

        System.out.println(result);
    }

    private static Credential getCredentials(NetHttpTransport httpTransport) throws IOException {
        // Load client secrets
        File initialFile = new File("credentials.json");
        InputStream targetStream = new FileInputStream(initialFile);
        GoogleClientSecrets clientSecrets = GoogleClientSecrets.load(GsonFactory.getDefaultInstance(), new InputStreamReader(targetStream));

        // Build flow and trigger user authorization request.
        GoogleAuthorizationCodeFlow flow = new GoogleAuthorizationCodeFlow.Builder(
                httpTransport, GsonFactory.getDefaultInstance(), clientSecrets,
                Collections.singletonList(SheetsScopes.SPREADSHEETS))
                .setDataStoreFactory(new FileDataStoreFactory(new File("tokens")))
                .setAccessType("offline")
                .build();

        LocalServerReceiver receiver = new LocalServerReceiver.Builder().setPort(8888).build();
        return new AuthorizationCodeInstalledApp(flow, receiver).authorize("user");
    }

    public static JsonNode HttpRequestData(String endPoint, String method, String payload, String info) {
        try {
            long timestamp = Instant.now().toEpochMilli();
            String signature = genSignature(payload, timestamp);

            Request request;
            if ("GET".equals(method)) {
                HttpUrl.Builder httpBuilder = Objects.requireNonNull(HttpUrl.parse(URL + endPoint)).newBuilder();
                httpBuilder.addQueryParameter("category", "linear");
                request = new Request.Builder()
                        .url(httpBuilder.build())
                        .addHeader("X-BAPI-API-KEY", API_KEY)
                        .addHeader("X-BAPI-SIGN", signature)
                        .addHeader("X-BAPI-SIGN-TYPE", "2")
                        .addHeader("X-BAPI-TIMESTAMP", String.valueOf(timestamp))
                        .addHeader("X-BAPI-RECV-WINDOW", "5000")
                        .addHeader("Content-Type", "application/json")
                        .build();
            } else {
                // Add POST logic here if needed
                return null;
            }

            Response response = httpClient.newCall(request).execute();
            if (response.body() != null) {
                String responseBody = response.body().string();
                System.out.println(responseBody);
                System.out.println(info + " Elapsed Time : " + response.receivedResponseAtMillis());

                JsonNode rootNode = objectMapper.readTree(responseBody);
                return rootNode.path("result").path("list");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String genSignature(String payload, long timestamp) throws Exception {
        String paramStr = timestamp + API_KEY + "5000" + payload;
        Mac sha256Hmac = Mac.getInstance("HmacSHA256");
        SecretKeySpec secretKey = new SecretKeySpec(SECRET_KEY.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        sha256Hmac.init(secretKey);
        byte[] hash = sha256Hmac.doFinal(paramStr.getBytes(StandardCharsets.UTF_8));
        StringBuilder hexString = new StringBuilder();
        for (byte b : hash) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) hexString.append('0');
            hexString.append(hex);
        }
        return hexString.toString();
    }
}
